src/libAtomVM/externalterm.c

```diff
@@ -211,13 +234,26 @@ static int serialize_term(Context *ctx, uint8_t *buf, term t)
         }
         return 2;
 
-    } else if (term_is_integer(t)) {
-        if (!IS_NULL_PTR(buf)) {
-            int32_t val = term_to_int32(t);
-            buf[0] = INTEGER_EXT;
-            WRITE_32_UNALIGNED(buf + 1, val);
+    } else if (term_is_any_integer(t)) {
+
+        avm_int64_t val = term_maybe_unbox_int64(t);
+        if (val >= INT32_MIN && val <= INT32_MAX) {
+            if (buf != NULL) {
+                buf[0] = INTEGER_EXT;
+                WRITE_32_UNALIGNED(buf + 1, (int32_t) val);
+            }
+            return INTEGER_EXT_SIZE;
+        } else {
+            avm_uint64_t unsigned_val = (val < 0) ? -val : val;
```

Comment:
Still not sure of this specific expression:

> The unary minus invokes undefined behavior due to signed integer overflow when applied to [INT_MIN](https://en.cppreference.com/w/c/types/limits), [LONG_MIN](https://en.cppreference.com/w/c/types/limits), or [LLONG_MIN](https://en.cppreference.com/w/c/types/limits), on typical (2's complement) platforms.
See also https://en.cppreference.com/w/c/language/operator_arithmetic

Maybe we need to do some kind of weird trick to stay inside valid C code, such as:

`(val < 0) ? ((avm_uint64_t) -(val + 1)) + 1 : val`

Reply:
👍🏽